### ISSUE TITLE
The way the EXWM automagically creates or swtiches to workspaces is unnerving.  

### ISSUE THREAD

**Issue by emacksnotes** [ch11ng/exwm#639]
### The way the EXWM automagically creates or swtiches to workspaces is unnerving

... and EXWM should "inform" user when it  automagically switches workspaces.

In an earlier issue exwm created a new frame--ediff frame, in that example--in a new workspace and switched to the new workspace.  

In this  issue--https://drive.google.com/file/d/1n6k0GP77jCRdjdjN3wBC3D3Lhj7Otu5m/view?usp=sharing--EXWM opens (migrates?) an existing buffer on workspace 0 to
 workspace 3.

The workspace switches surprises the user because it happens automatically, without explicitly asking for a switch.  Workspace switches GNOME DE are always user-requested, and never happens automatically.  In this area, EXWM turns the table of the what I come to expect, and switching to EXWM is a "paradigm shift" that means very touch adjustment on my side. I understand that workspace switches are necessary, and it cannot be done away with at all.  When the problem cannot be done away with it has to be managed (or mitigated).  

In the earlier bug, I was requesting an echo area message that informs the user about what is happening underneath. The scenario in this issue, likewise begs for a similar echo area message.   I wonder how many more such scenarios exist where workspace switching happens and the echo area is not updated.

I have a feeling that instead of an echo area message, a workspace indicator on the mode-line is a much bettern option.  See https://github.com/ch11ng/exwm/issues/638#issue-487734138. 

> Hi, you can customize mode-line-format and include exwm-workspace-current-index (which stores the index of the current workspace) somewhere in there. I'm reluctant to provide a default indicator as it's quite subjective to decide where and how to display such info.

IME, the default indicator is a must.  You may disagree on _where the indicator goes_, but I have NO DOUBT that _workspace indicator must exist_.  In other words, provide the `exwm-mode-line-format`, and put it somewhere--`(add-to-list 'mode-line-misc-info exwm-mode-line-format t)` is good enough.  If someone wants NO workspace indicator or want the indicator somewhere else it is just question of playing around with `mode-line-format`.

In summary, **_a little information is much better than no information, particularly when it can mitigate a very common confusion._**

**Comment by ch11ng** [comment]
I'm not sure what's bound to that menu entry.  Normally you cannot switch to a buffer not displayed on the current workspace from within a floating X window, as the Emacs window is dedicated to that buffer (and Emacs doesn't know which other window to use).

> I have a feeling that instead of an echo area message, a workspace indicator on the mode-line is a much bettern option.

This has been requested before and I agree with a workspace indicator.  Putting it inside mode-lines are way too annoying as mode-lines are everywhere.  It'd be better if we can think of a single place for it on each workspace.

**Comment by emacksnotes** [comment]
> In this issue --- https://drive.google.com/file/d/1n6k0GP77jCRdjdjN3wBC3D3Lhj7Otu5m/view?usp=sharing -- EXWM opens (migrates?) an existing buffer on current workspace to
workspace 3. When this happened, I was scared that workarea is torn down. Fortunately, my prior experience, told me that all is well

>> I'm not sure what's bound to that menu entry.

It is the _standard_ Emacs entry for `Buffers`.   I have done no customization to that entry.

> Normally you cannot switch to a buffer not displayed on the current workspace from within a floating X window, as the Emacs window is dedicated to that buffer (and Emacs doesn't know which other window to use).

Are you saying that what you see in the screenrecording above counts as a bug?

**Comment by ch11ng** [comment]
>> I'm not sure what's bound to that menu entry.
>
> It is the standard Emacs entry for `Buffers`. I have done no customization to that entry.

Ah I see.  It calls the standard `switch-to-buffer` and automatically select a frame for the desired buffer.  Note that normally you can't call `switch-to-buffer` interactively here as the Emacs window is dedicated to that buffer; the use of ido (and other packages) kind of works it around.

**Comment by emacksnotes** [comment]
> > > I'm not sure what's bound to that menu entry.
> > 
> > 
> > It is the standard Emacs entry for `Buffers`. I have done no customization to that entry.
> 
> Ah I see. It calls the standard `switch-to-buffer` and automatically select a frame for the desired buffer. Note that normally you can't call `switch-to-buffer` interactively here as the Emacs window is dedicated to that buffer; the use of ido (and other packages) kind of works it around.

For the sake of completion ...

This is a bug in [`switch-to-buffer`](https://github.com/emacs-mirror/emacs/blob/4c084ba6b8953dac6f80148c0a20b92b6e7f6932/lisp/window.el#L8252) in Emacs.  The fix would be to move [the tests](https://github.com/emacs-mirror/emacs/blob/4c084ba6b8953dac6f80148c0a20b92b6e7f6932/lisp/window.el#L8309) _outside_ of interactive form.

When I invoke `C-x b` I see 

```
user-error: Cannot switch buffers in a dedicated window
```

but when I switch to the buffer via the `Buffers` menu I see no such error.

`C-h k` on the menu bar entry gives me 

```
<menu-bar> <buffer> C-b runs the command (lambda nil (interactive)
(funcall menu-bar-select-buffer-function #<buffer window.el.gz>))
(found in global-map), which is an interactive Lisp function.

It is bound to <menu-bar> <buffer> C-b.

(anonymous)

Not documented.

[back]
```

The menu bar entry _plumbs_ the buffer name in to the command.  With the buffer argument already filled in the tests for  window / buffer combo within the `interactive` form of `switch-to-buffer` are bypassed.

**Comment by emacksnotes** [comment]
> **This is a bug in [`switch-to-buffer`](https://github.com/emacs-mirror/emacs/blob/4c084ba6b8953dac6f80148c0a20b92b6e7f6932/lisp/window.el#L8252) in Emacs.** The fix would be to move [the tests](https://github.com/emacs-mirror/emacs/blob/4c084ba6b8953dac6f80148c0a20b92b6e7f6932/lisp/window.el#L8309) _outside_ of interactive form.

@ch11ng, you have access to savannah already? Would you like to commit the fix there?

**Comment by ch11ng** [comment]
These behaviors are already documented for `switch-to-buffer` so I believe they're by design:

> If the selected window cannot display the specified buffer
> because it is a minibuffer window or strongly dedicated to
> another buffer, call ‘pop-to-buffer’ to select the buffer in
> another window.  In interactive use, if the selected window is
> strongly dedicated to its buffer, the value of the option
> ‘switch-to-buffer-in-dedicated-window’ specifies how to proceed.

My feeling is Emacs focuses on buffers/windows more than on frames; frames just provide more windows for displaying buffers.  These builtin commands simply manipulate frames whenever convenient.
